### PR TITLE
Add retry middleware to sync committee prover

### PR DESCRIPTION
**Pull request by Wizdave97**
Enable retries when fetching sync committee data

### CODE REVIEW

File: modules/consensus/sync-committee/prover/src/middleware.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        };
+        match result {
+            Ok(response) => {
+                if response.status() == StatusCode::OK {
+                    return Ok(Some(response));
+                };
+                let next_index = next_state()?;
+                _state.active_url_index = next_index;
+                let url = self.providers[next_index].clone();
+                request.headers_mut().insert(HOST, url.parse().map_err(|e| anyhow!("{e:?}"))?);
```
Comment (seunlanlege) [inline modules/consensus/sync-committee/prover/src/middleware.rs:48]: would be good to get tests for this, unsure if this is correct. I think you want to modify `request.url`

Comment (Wizdave97) [reply modules/consensus/sync-committee/prover/src/middleware.rs:48]: I'll add tests some tests.

**Comment by Wizdave97** [comment]
Should we integrate in this middleware instead of the regular retry?

File: modules/consensus/sync-committee/prover/src/test.rs [hunk 2]
```diff
@@ -263,6 +267,28 @@ async fn test_prover() {
     }
 }
 
+#[tokio::test]
```
[the remark this answers is not in the document]
Comment (Wizdave97) [reply modules/consensus/sync-committee/prover/src/test.rs:270]: Yeah

**Comment by Wizdave97** [comment]
> > Should we integrate in this middleware instead of the regular retry?
> 
> Yeah i don't think we need `reqwest-retry` we just need the provider switching middleware. Although now that i think about it, why not both? I'm assuming if all urls fails, `reqwest-retry` would kick in?

Yeah but it would only retry on the last provider url.

**Comment by Wizdave97** [comment]
Or we use retry first, then switch as the second strategy.

**Comment by seunlanlege** [comment]
> Or we use retry first, then switch as the second strategy.

I think instead, we cycle back to the first url?

**Comment by Wizdave97** [comment]
> I think instead, we cycle back to the first url?

We can only retry a maximum of the number of urls + the number of retries configured in the `reqwest-retry` middleware.
we can't switch the urls used by the `reqwest-retry` middleware.

 Do you mean we should just modify the request to use the first url before exiting the provider switcher?

**Review by seunlanlege — APPROVED** [review]

**Pull request merged**